Enable raytracer smoother for SymmetricCylinder - work in progress

Lathreas opened the pull request:
This contains a draft to make raytraced surfaces smooth near the boundaries using an arbitrary smooth falloff function. Currently the corners are not yet implemented, to be done shortly. Code cleanup will also be done.

Lathreas committed:
Fix calculation of t in SymmetricCylinder.RayIntersect()

The calculation of t assumed that the Ray was not rotated, whereas
it actually was rotated. Therefore, it returned the wrong t. This
patch fixes this by assuming the Ray was rotated, which actually
simplifies the calculation.

Lathreas committed:
Add RaySurfaceIntersection struct and switch output of RayIntersect()

Lathreas committed:
Add IExtendedRaytraceableSurface

Lathreas committed:
Add barebones SmoothMoldCastMap framework

Lathreas committed:
Implement exponential smoothing function

Lathreas committed:
Add deformed plane raytracer

This will be useful for smoothing the edges of the cylinder raytracer

Change raytracer smoothing weight to sigmoidal curve

Add plane extension to SymmetricCylinder for smoothing and fix bug in Plane

AdamNorberg commented on engine/geometry/IExtendedRaytraceableSurface.cs:
```diff
@@ -0,0 +1,60 @@
+		/// </summary>
+		public float DistanceToSurface { get; set; }
+
+		public RayExtendedSurfaceIntersection(float rayLength, float distanceToSurface)
```
This constructor doesn't pull its weight - I recommend removing it. C# allows struct literal initialization for structs with only public properties/fields, and then it's immediately obvious at the call site that nothing subtle is going on (because it uses brace initialization instead of invoking a constructor with parentheses).

(Note that this is different from C++, where - courtesy of initializer_list constructors - there is absolutely no guarantee that brace initialization is any less confusing than standard constructor initialization...)

AdamNorberg commented on engine/geometry/IExtendedRaytraceableSurface.cs:
```diff
@@ -0,0 +1,60 @@
+	public struct RayExtendedSurfaceIntersection
+	{
+		/// <summary>
+		/// 	The distance from the start point of the ray to the intersection point.
```
What value does this hold if the ray does not intersect the surface?

AdamNorberg commented on engine/geometry/IExtendedRaytraceableSurface.cs:
```diff
@@ -0,0 +1,60 @@
+	/// <summary>
+	/// 	An interface defining the ability to raytrace a surface analytically.
+	/// </summary>
+	public struct RayExtendedSurfaceIntersection
```
It seems odd for this "Intersection" struct not to actually declare where (if anywhere) the intersection occurs, like RaySurfaceIntersection does.  I would kind of expect a RayExtendedSurfaceIntersection to have a RaySurfaceIntersection field and its extra field(s), and add properties to "call through" to the public properties of the embedded RaySurfaceIntersection, although this would complicate construction (so my comment below about just using brace initialization would be wrong).

AdamNorberg commented on engine/geometry/IRaytraceableSurface.cs:
```diff
@@ -16,6 +16,34 @@
+		/// </summary>
+		public float V { get; set; }
+
+		public RaySurfaceIntersection(float rayLength, float u, float v)
```
Again, this is a good candidate for removal, brace initialization will do.

AdamNorberg commented on engine/geometry/IRaytraceableSurface.cs:
```diff
@@ -16,6 +16,34 @@
+	/// <summary>
+	/// 	The intersection point on a surface after a raycast.
+	/// </summary>
+	public struct RaySurfaceIntersection
```
For all fields: What values does this hold if it represents a non-intersection?

AdamNorberg commented on engine/geometry/Plane.cs:
```diff
@@ -0,0 +1,157 @@
+	/// 	A <c>Plane</c> represents a planar surface that can be deformed by a heightmap. Due to its simplicity, it is
+	/// 	raytraceable.
+	/// </summary>
+	public class Plane : IRaytraceableSurface
```
I don't like the name Plane for this, since it is not a geometric plane due to its use of heightmap deformation. Consider HeightmappedPlane, which seems more obvious to me.

AdamNorberg commented on engine/geometry/Plane.cs:
```diff
@@ -0,0 +1,157 @@
+		/// 	The heightmap that deforms the plane in the normal direction. The normal is defined as the normalized
+		/// 	cross product between <c>VectorU</c> and <c>VectorV</c>.
+		/// </param>
+		public Plane(Vector3 origin, Vector3 vectorU, Vector3 vectorV, RaytraceableFunction1D heightmap, float heightFactor = 1.0f)
```
I feel like this is a spot where explicit objects for "vector spaces" would reduce the opportunity for bugs, fwiw. (going back to that CL where I drafted the idea out, which is still just kinda sitting.)

More immediately practical: what happens if a Plane is not uniquely defined by the provided arguments? If vectorU is some multiple of vectorV, or either is zero (which, I suppose, is just the zero multiple...), no plane is defined.

AdamNorberg commented on engine/geometry/Plane.cs:
```diff
@@ -0,0 +1,157 @@
+
+		#region Properties
+		// TODO: change access modifier
+		protected RaytraceableFunction1D heightmap1D;
```
That the plane is only deformable from a 1d argument rather than 2d surprises me. It's exactly what the constructor documentation says and what types are involved, so I don't think there's anything to immediately change, I'm just surprised and I'm interested in why you're using a 1d map here.

Lathreas replied:
It's meant to be a 1D deformation to give a corrugated effect, so that whatever shape the edge of a HingeJoint is becomes extended and smoothly falls off (if that makes sense, otherwise I can explain the method with a few pictures - probably good idea anyway for documentation purposes). Indeed, from the name it feels like it should be 2D, so that is indeed confusing. I've changed the name accordingly (see comment above).

AdamNorberg commented on engine/geometry/Plane.cs:
```diff
@@ -0,0 +1,157 @@
+				Matrix4x4 planeMatrix = new Matrix4x4(VectorU.X, VectorU.Y, VectorU.Z, 0.0f,
+													  VectorV.X, VectorV.Y, VectorV.Z, 0.0f,
+													  Normal.X,  Normal.Y,  Normal.Z,  0.0f,
+													  0.0f,      0.0f,      0.0f,      1.0f);
```
Here and everywhere you're constructing these planeMatrix objects, something has gone wrong with indentation. Are you indenting with tabs, using a tab stop size different from Github's expectation?

Either indent with spaces or indent _only_ with tabs, including using them as column breaks, and break after the opening parenthesis to avoid needing to try to align tabbed continuation rows with the opening parenthesis (which is in a character-fixed position).

FWIW, I prefer tab indentation _when it is consistently and correctly used_, which means strictly forbidding the "first line of arguments inline with opening paren, subsequent lines indented straight under it" pattern because it's impossible to do with varied tab stop sizes. My reason for preferring it is because it's useful for editors that use proportional fonts and use font size as part of syntax highlighting - I think Source Insight is the only popular one, admittedly, but I wish it was more common.

tl;dr: indentation broken, plz fix.

AdamNorberg commented on engine/geometry/Plane.cs:
```diff
@@ -0,0 +1,157 @@
+				// of this region, the ray does not intersect the cylinder.
+				if (!(u >= 0.0f && u <= 1.0f) || !(v >= 0.0f && v <= 1.0f))
+				{
+					break;
```
Is the output of heightmap1D.SolveRaytracePlanar sorted? If so, why continue iteration after finding anything that works, since anything later in the list will be further away? If not, why is this `break` rather than `continue`?

Lathreas replied:
Oh, good catch! Should be `continue` indeed. Fixed.

AdamNorberg commented on engine/geometry/SymmetricCylinder.cs:
```diff
@@ -168,24 +168,25 @@ public float RayIntersect(Ray ray)
+				// of this region, the ray does not intersect the cylinder.
+				if (!(t >= 0.0f && t <= 1.0f))
+				{
+					break;
```
As above: why not `continue`?

AdamNorberg commented on engine/geometry/SymmetricCylinder.cs:
```diff
@@ -196,12 +197,89 @@ public float RayIntersect(Ray ray)
 				if ( phi > StartAngle.GetValueAt(t) && phi < EndAngle.GetValueAt(t) && i >= 0.0f)
 				{
-					minimum = Math.Sign(i)*(float)Math.Min(Math.Abs(minimum), Math.Abs(i));
+					if (Math.Abs(i) < Math.Abs(minimum.RayLength))
```
As above, this check should be reversed and first.

AdamNorberg commented on engine/geometry/Plane.cs:
```diff
@@ -0,0 +1,157 @@
+				// otherwise discard:
+				if (i >= 0.0f)
+				{
+					if (Math.Abs(i) < Math.Abs(minimum.RayLength))
```
If the output isn't sorted, I feel like the inversed version of this check should be first and call `continue` if it's not met, since it's a cheap float vs. float comparison that filters whether this iteration can possibly do anything (we won't have a new minimum if this intersection is further away, after all).

AdamNorberg replied:
Additionally, the Math.Abs(i) call is not required since we already know that i is mostly nonnegative. (It may be -0.0 but nothing bad happens if it's negative zero so the existing check is "nonnegative enough".)

AdamNorberg commented on engine/geometry/Plane.cs:
```diff
@@ -0,0 +1,157 @@
+					break;
+				}
+
+				// Determine if the ray is inside the pie-slice of the cylinder that is being displayed,
```
What do cylinders have to do with this? (I infer this is a bad copy-paste from the Cylinder versions of this logic)

AdamNorberg commented on engine/geometry/SymmetricCylinder.cs:
```diff
@@ -196,12 +197,89 @@ public float RayIntersect(Ray ray)
+			{
+				return new RayExtendedSurfaceIntersection(intersection.RayLength, 0.0f);
+			}
+			else
```
This is an unpleasantly wedge-shaped function! When an if block ends in a `return` on all codepaths (as a one-line return block clearly does!), skip the `else` - the only way execution continues is implicitly `else`, so you can get rid of a lot of indentation by just removing all the `else`s subsequent to `if...return` blocks. (that is, remove the else, remove the curly brackets, outdent the code so it runs "unconditionally" - except execution of the function will already have halted if the `if` was taken.)

Lathreas committed:
Fix small numerical errors using Newton-Raphson

The small remaining numerical errors caused by using lower-order
approximations can be fixed by approaching the final solution
iteratively using the Newton-Raphson method. This currently
doesn't have a noticeable performance impact, and solves most
visual artifacts. Of course, we can later optimize this further.

Lathreas committed:
Create dotnet.yml workflow

This sets up continuous integration.

Lathreas committed:
Create godot-ci.yml

Lathreas committed:
Update godot-ci.yml

Lathreas committed:
test

Lathreas committed:
Update godot-ci.yml

Lathreas committed:
Create godot-ci.yml

Lathreas committed:
Create godot-ci.yml

Lathreas committed:
Add export presets

Lathreas committed:
Fix .csproj files for exporting project

Lathreas committed:
Lower ExampleBone vertex resolution for webexport

The default vertex count made the program way too slow.

Lathreas committed:
Add Isosurface class - not working yet

Lathreas committed:
Enable smooth raytracing for cup-shaped bone as well.

Not finished yet though.

Lathreas committed:
Merge branch 'master' into smooth-raytracer

Lathreas committed:
Fix build chain

Lathreas committed:
Allow rotating bone with Ctrl+Scrollwheel
